File: src/main/java/no/novari/flyt/mapping/service/ValueMappingService.java
package no.novari.flyt.mapping.service;


import no.novari.flyt.mapping.model.configuration.ValueMapping;
import no.novari.flyt.mapping.model.instance.InstanceObject;
import org.springframework.stereotype.Service;

import java.util.Map;
import java.util.Objects;

@Service
public class ValueMappingService {

    private final InstanceReferenceService instanceReferenceService;
    private final ValueConvertingService valueConvertingService;

    public ValueMappingService(
            InstanceReferenceService instanceReferenceService,
            ValueConvertingService valueConvertingService
    ) {
        this.instanceReferenceService = instanceReferenceService;
        this.valueConvertingService = valueConvertingService;
    }

    public Object toValue(
            ValueMapping valueMapping,
            Map<String, String> instanceValuePerKey,
            InstanceObject[] selectedCollectionObjectsPerKey
    ) {
        return Objects.isNull(valueMapping.getMappingString()) ? null : switch (valueMapping.getType()) {
            case BOOLEAN -> Boolean.parseBoolean(valueMapping.getMappingString());
            case STRING, URL -> valueMapping.getMappingString();
            case FILE, DYNAMIC_STRING -> instanceReferenceService.replaceIfReferencesWithInstanceValues(
                    valueMapping.getMappingString(),
                    instanceValuePerKey,
                    selectedCollectionObjectsPerKey
            );
            case VALUE_CONVERTING -> valueConvertingService.convertValue(
                    valueMapping.getMappingString(),
                    instanceValuePerKey,
                    selectedCollectionObjectsPerKey
            );
        };
    }

}

File: src/main/kotlin/no/novari/flyt/mapping/service/ValueMappingService.kt
package no.novari.flyt.mapping.service

import no.novari.flyt.mapping.model.configuration.ValueMapping
import no.novari.flyt.mapping.model.instance.InstanceObject
import org.springframework.stereotype.Service

@Service
class ValueMappingService(
    private val instanceReferenceService: InstanceReferenceService,
    private val valueConvertingService: ValueConvertingService,
) {
    fun toValue(
        valueMapping: ValueMapping,
        instanceValuePerKey: Map<String, String?>,
        selectedCollectionObjectsPerKey: Array<InstanceObject>,
    ): Any? {
        val mappingString = valueMapping.mappingString ?: return null

        return when (valueMapping.type) {
            ValueMapping.Type.BOOLEAN -> {
                mappingString.toBoolean()
            }

            ValueMapping.Type.STRING,
            ValueMapping.Type.URL,
            -> {
                mappingString
            }

            ValueMapping.Type.FILE,
            ValueMapping.Type.DYNAMIC_STRING,
            -> {
                instanceReferenceService.replaceIfReferencesWithInstanceValues(
                    mappingString,
                    instanceValuePerKey,
                    selectedCollectionObjectsPerKey,
                )
            }

            ValueMapping.Type.VALUE_CONVERTING -> {
                valueConvertingService.convertValue(
                    mappingString,
                    instanceValuePerKey,
                    selectedCollectionObjectsPerKey,
                )
            }
        }
    }
}
